common/src/primitives/merkle/tree/tree_size.rs

```diff
@@ -0,0 +1,230 @@
+    HugeTreeUnsupported(usize),
+}
+
+impl TreeSize {
+    pub fn get(&self) -> usize {
+        self.0
+    }
+
+    pub fn leaf_count(&self) -> NonZeroUsize {
+        ((self.0 + 1) / 2).try_into().expect("Guaranteed by construction")
+    }
+
+    pub fn level_count(&self) -> NonZeroUsize {
+        (self.0.count_ones() as usize).try_into().expect("Guaranteed by construction")
+    }
+
+    pub fn from_value(value: usize) -> Result<Self, TreeSizeError> {
+        Self::try_from(value)
+    }
+
+    pub fn from_leaf_count(leaf_count: usize) -> Result<Self, TreeSizeError> {
+        if leaf_count == 0 {
+            return Err(TreeSizeError::ZeroSize);
+        }
+        Self::try_from(leaf_count * 2 - 1)
+    }
+}
+
+impl TryFrom<usize> for TreeSize {
+    type Error = TreeSizeError;
+
+    fn try_from(value: usize) -> Result<Self, Self::Error> {
+        if value == 0 {
+            Err(TreeSizeError::ZeroSize)
+        } else if (value + 1).count_ones() != 1 {
+            Err(TreeSizeError::InvalidSize(value))
+        } else if value > MAX_TREE_SIZE {
+            Err(TreeSizeError::HugeTreeUnsupported(value))
+        } else {
+            Ok(Self(value))
+        }
+    }
+}
+
+impl From<TreeSize> for usize {
+    fn from(tree_size: TreeSize) -> Self {
+        tree_size.0
+    }
+}
+
+impl AsRef<usize> for TreeSize {
+    fn as_ref(&self) -> &usize {
+        &self.0
+    }
+}
+
+impl Display for TreeSize {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn construction_from_tree_size() {
+        assert_eq!(TreeSize::try_from(0), Err(TreeSizeError::ZeroSize));
+        assert_eq!(TreeSize::try_from(1), Ok(TreeSize(1)));
+        assert_eq!(TreeSize::try_from(2), Err(TreeSizeError::InvalidSize(2)));
+        assert_eq!(TreeSize::try_from(3), Ok(TreeSize(3)));
+        assert_eq!(TreeSize::try_from(4), Err(TreeSizeError::InvalidSize(4)));
+        assert_eq!(TreeSize::try_from(5), Err(TreeSizeError::InvalidSize(5)));
+        assert_eq!(TreeSize::try_from(6), Err(TreeSizeError::InvalidSize(6)));
+        assert_eq!(TreeSize::try_from(7), Ok(TreeSize(7)));
+        assert_eq!(TreeSize::try_from(8), Err(TreeSizeError::InvalidSize(8)));
+        assert_eq!(TreeSize::try_from(9), Err(TreeSizeError::InvalidSize(9)));
+        assert_eq!(TreeSize::try_from(10), Err(TreeSizeError::InvalidSize(10)));
+        assert_eq!(TreeSize::try_from(11), Err(TreeSizeError::InvalidSize(11)));
+        assert_eq!(TreeSize::try_from(12), Err(TreeSizeError::InvalidSize(12)));
+        assert_eq!(TreeSize::try_from(13), Err(TreeSizeError::InvalidSize(13)));
+        assert_eq!(TreeSize::try_from(14), Err(TreeSizeError::InvalidSize(14)));
+        assert_eq!(TreeSize::try_from(15), Ok(TreeSize(15)));
+        assert_eq!(TreeSize::try_from(16), Err(TreeSizeError::InvalidSize(16)));
+
+        for i in 1..1000usize {
```

Comment:
"proptest" as a library, I'm not gonna do that, because it's a macro that prevents formatting.

But you're right. Random values are more appropriate. I'm going for that.

Reply:
I unresolved this, and wrote the exhaustive cases. Take a look.

Reply:
Looks good